error.go

```diff
@@ -475,3 +493,114 @@ func AppendInto(into *error, err error) (errored bool) {
+// but defer the invocation of scanner.Err() until the function returns.
+//
+// 	defer multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))
+type Invoke func() error
+
+// Invoke calls the supplied function and returns its result.
+func (f Invoke) Invoke() error { return f() }
+
+// Close builds an Invoker that closes the provided io.Closer. Use it with
+// AppendInvoke to close io.Closers and append their results into an error.
+//
+// For example,
+//
+// 	func processFile(path string) (err error) {
+// 		f, err := os.Open(path)
+// 		if err != nil {
+// 			return err
+// 		}
+// 		defer multierr.AppendInvoke(&err, multierr.Close(f))
+// 		return processReader(f)
+// 	}
+//
+// In this example, multierr.Close will construct the Invoker right away, but
+// defer the invocation of f.Close until the function returns.
+//
+// 	defer multierr.AppendInvoke(&err, multierr.Close(f))
+func Close(closer io.Closer) Invoker {
+	return Invoke(closer.Close)
+}
+
+// AppendInvoke appends the result of calling the given Invoker into the
+// provided error pointer. Use it with named returns to safely defer
+// invocation of fallible operations until a function returns, and capture the
+// resulting errors.
+//
+// 	func doSomething(...) (err error) {
+// 		// ...
+// 		f, err := openFile(..)
+// 		if err != nil {
+// 			return err
+// 		}
+//
+// 		// multierr will call f.Close() when this function returns and
+// 		// if the operation fails, its append its error into the
+// 		// returned error.
+// 		defer multierr.AppendInvoke(&err, multierr.Close(f))
+//
+// 		scanner := bufio.NewScanner(f)
+// 		// Similarly, this scheduled scanner.Err to be called and
+// 		// inspected when the function returns and append its error
+// 		// into the returned error.
+// 		defer multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))
+//
+// 		// ...
+// 	}
+//
+// Without deferrals, AppendInvoke behaves exactly like AppendInto.
+//
+// 	err := // ...
+// 	multierr.AppendInvoke(&err, mutltierr.Invoke(foo))
+//
+// 	// ...is roughly equivalent to...
+//
+// 	err := // ...
+// 	multierr.AppendInto(&err, foo())
+//
+// The advantage of the indirection introduced by Invoker is to make it easy
+// to defer the invocation of a function. Without this indirection, the
+// invoked function will be evaluated at the time of the deferral rather than
+// when the function returns.
+//
+// 	defer multierr.AppendInto(&err, foo())
+// 		// BAD: This is likely not what the caller intended. This will
+// 		// evaluate foo() right away and append its result into the
+// 		// error when the function returns.
+//
+// 	defer multierr.AppendInvoke(&err, multierr.Invoke(foo))
+// 		// GOOD: This will defer invocation of foo unutil the function
+// 		// returns.
+//
+// multierr provides a few Invoker implementations out of the box for
+// convenience. See Invoker for more information.
+func AppendInvoke(into *error, call Invoker) {
+	AppendInto(into, call.Invoke())
```

Comment:
I was already thinking about this at the time implementing it. Back then I came to the conclusion that the `bool` return might be more confusing because e.g. in the case you mentioned it would rather be

```go
if multierr.AppendInvoke(&err, multierr.Close(f)) {
    // work work
}
```

and I'm not sure what's the advantage of this over

```go
if multierr.AppendInto(&err, f.Close()) {
    // work work
}
```

looking at both I still prefer the current implementation to not confuse users of the library about when to use which method if both are suited for some overlapping use cases but might lead to very unexpected issues when used wrong :thinking:

Reply:
Yeah, I thought about whether AppendInvoke should return something as well and came to the conclusion that for the cases where you want to operate on the return value of AppendInvoke, you'd be better off using just AppendInto. @prashantv WDYT?

Reply:
Agree, there doesn't seem to be much of a benefit over `AppendInto`, so let's skip for now.